### DISCUSSION TITLE
Allowance, approval and permit hooks

### DISCUSSION THREAD

**Discussion by smakosh** [wevm/wagmi#1036] [Ideas]
Hi everyone, this is my first discussion on this repository, really grateful for what you've built here!

As I was working on implementing allowance, approvals and permits, I found myself re-using lots of useRead/WriteContract hooks and I've been thinking why not provide hooks to make this much easier?

Example:
![image](https://user-images.githubusercontent.com/20082141/193583340-819da0a0-c3d0-42f0-925a-5a3f5d9ae641.png)

Happy to work on this and raise a PR if this seems interesting to the community.

**Comment by sarinniural** [comment]
@smakosh do you have an implementation for this you can share ?

**Reply by smakosh** [reply]
It is on Diagonal demo, should be open sourced soon. Will have something ready before that.

**Reply by sarinniural** [reply]
@smakosh ok I will be waiting 😅

**Reply by nfts2me** [reply]
@smakosh I am also interested in these hooks!

**Reply by smakosh** [reply]
First preview of a working `useAllowance` hook 
![image](https://user-images.githubusercontent.com/20082141/197091401-6b106fd3-d3eb-4ebb-8bc0-524736ca8e41.png)

**Reply by Markkos89** [reply]
looking forward to this!

**Comment by smakosh** [comment]
Initial PR that adds `useAllowance` here: https://github.com/wagmi-dev/wagmi/pull/1136

More hooks to be followed.

**Comment by smakosh** [comment]
thinking of having these hooks in a fork of wagmi over at: https://github.com/smakosh/wagmi-hooks, as there's no intent to have these hooks in wagmi but to provide the functionality through a CLI gen.

**Reply by tmm** [reply]
Sounds great. You are welcome to add those to [awesome-wagmi](https://github.com/wagmi-dev/awesome-wagmi)!

**Comment by Willgg** [comment]
ERC20Permit (eip2612) is a great way to avoid the Approve + Transfer pattern which requires too much gas so having something to trigger the signature + call the contract transfer function would be great !

**Comment by RastogiAbhijeet** [comment]
https://github.com/dragonfly-xyz/useful-solidity-patterns/tree/main/patterns/permit2
It might make sense to also integrate permit2 if not already

**Reply by DeveloperTheExplorer** [reply]
YES! After moving to Wagmi on Viem (instead of ethers) now it seems that there is no way to use the UniSwap Permit2 SDK...

**Comment by vacekj** [comment]
Made a small library for handling permits with wagmi and viem here: https://github.com/vacekj/wagmi-permit/

**Reply by kryptopaul** [reply]
Thanks! I was looking for something exactly like this :)

**Comment by raffle-x-dev** [comment]
Hey is this still a thing? Looks amazing.

**Reply by tmm** [reply]
You can use the [Wagmi CLI](https://wagmi.sh/cli/getting-started) to generate hooks/actions like this.

**Reply by raffle-x-dev** [reply]
> You can use the [Wagmi CLI](https://wagmi.sh/cli/getting-started) to generate hooks/actions like this.

I just installed the CLI and tried to run the example, but I get this error:
```
npx wagmi generate

✔ Validating plugins
⠧ Resolving contracts
Invalid response
- Invalid literal value, expected "OK" at 'message'
```

any ideas?

**Reply by tmm** [reply]
Hard to tell without seeing your Wagmi CLI config file. Are you using the Etherscan plugin? Have you added the [`apiKey`](https://wagmi.sh/cli/api/plugins/etherscan#apikey) property?

**Reply by DobromirKirovLime** [reply]
what is the state of this...?

**Comment by based64-eth** [comment]
why arent more people asking about this?
